driver/src/price_estimation/dexag/api.rs

```diff
@@ -0,0 +1,183 @@
+#![allow(dead_code)]
+
+use anyhow::{anyhow, Context, Result};
+use ethcontract::Address;
+use isahc::prelude::*;
+use serde::{de::Error as _, Deserialize, Deserializer};
+use serde_with::rust::display_fromstr;
+use std::time::Duration;
+use url::Url;
+
+#[cfg_attr(test, mockall::automock)]
+/// Parts of the dex.ag api.
+pub trait DexagApi {
+    /// https://docs.dex.ag/api/tokens
+    fn token_list_full(&self) -> Result<Vec<Token>>;
+    /// https://docs.dex.ag/api/price
+    fn price(
+        &self,
+        from: &str,
+        to: &str,
+        amount: EitherAmount,
+        dex: Dex,
+        discluded: &str,
+    ) -> Result<Price>;
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct Token {
+    pub name: String,
+    pub symbol: String,
+    pub address: Option<Address_>,
```

Comment:
I'm not a big fan of the wrapper type given that it would have to be publicly exposed. Can we not implement custom deserialisation for this field?

Reply:
Not a fan either. If this was a normal `Address` and not an `Option` then we could use the serde `deserialize_with` attribute but I don't think it works in the option case so as far as I can tell we would have to manually implement `Deserialize` for the whole token struct which I dislike even more than exposing the wrapper.

Reply:
The other way is to have two structs. One for the public trait api which doesn't implement deserialize and has the correct `Address` type. The other private struct has `address: String` and does derive Deserialize and then we convert between the structs in the function.
I like this idea. Will implement and see how it feels when its written down.

Reply:
Can we not use deserialize_with on Option<Address> ?

Reply:
I expected this to be more complex but it turns out simple.